brush-core/src/interp.rs

```diff
@@ -109,32 +109,45 @@ struct PipelineExecutionContext<'a> {
 #[derive(Clone, Default)]
 pub struct ExecutionParameters {
     /// The open files tracked by the current context.
-    pub open_files: openfiles::OpenFiles,
+    pub(crate) open_files: openfiles::OpenFiles,
     /// Policy for how to manage spawned external processes.
     pub process_group_policy: ProcessGroupPolicy,
 }
 
 impl ExecutionParameters {
     /// Returns the standard input file; usable with `write!` et al.
-    pub fn stdin(&self) -> openfiles::OpenFile {
+    pub fn stdin(&self) -> impl std::io::Read {
         self.fd(0).unwrap()
     }
 
     /// Returns the standard output file; usable with `write!` et al.
-    pub fn stdout(&self) -> openfiles::OpenFile {
+    pub fn stdout(&self) -> impl std::io::Write {
         self.fd(1).unwrap()
     }
 
     /// Returns the standard error file; usable with `write!` et al.
-    pub fn stderr(&self) -> openfiles::OpenFile {
+    pub fn stderr(&self) -> impl std::io::Write {
         self.fd(2).unwrap()
     }
 
     /// Returns the file descriptor with the given number.
     #[allow(clippy::unwrap_in_result)]
-    pub fn fd(&self, fd: u32) -> Option<openfiles::OpenFile> {
+    pub(crate) fn fd(&self, fd: u32) -> Option<openfiles::OpenFile> {
         self.open_files.files.get(&fd).map(|f| f.try_dup().unwrap())
     }
+
+    pub(crate) fn stdin_file(&self) -> openfiles::OpenFile {
```

Comment:
I know it's kinda late but shouldn't this be:

```suggestion
    pub(crate) fn stdin_file(&self) -> impl std::io::Write {
```

Reply:
Fair question. I'd prefer not to have this fn at all. The change updated `stdin()` et al. to return an `impl` of a trait in `std::io`. Almost all call sites were perfectly fine with that because they were mostly just passing the value to `write!` or similar. 

But... there were a few internal-to-brush-core callers that were doing something more complicated. Rather than sort them out more properly I opted to add a few `pub(crate)` helpers (e.g. stdin_file()` that still returned an `OpenFile` or similar.

In the medium term, I'd like to take a pass over `pub(crate)` functions on exported types (like these) and find a way to remove them or properly export them.

And thanks for taking a look, @Elsie19 🙂 I expedited this one in before the end of the weekend, but am always happy to get the feedback whether it's before or after merge. I'm also happy to let changes like this sit for a bit in the future if I know there's interest / willingness for someone else to read through them.